acurl/acurl/utils.py

```diff
@@ -0,0 +1,54 @@
+from collections import defaultdict
+
+
+class CaseInsensitiveDefaultDict(defaultdict):
+    # Base on https://stackoverflow.com/a/32888599 but tweaked for python3
+    # and using defaultdict as a base
+    @classmethod
+    def _k(cls, key):
+        return key.lower() if isinstance(key, str) else key
+
+    def __init__(self, *args, **kwargs):
+        super(CaseInsensitiveDefaultDict, self).__init__(*args, **kwargs)
+        self._convert_keys()
+
+    def __getitem__(self, key):
+        return super(CaseInsensitiveDefaultDict, self).__getitem__(self.__class__._k(key))
+
+    def __setitem__(self, key, value):
+        super(CaseInsensitiveDefaultDict, self).__setitem__(self.__class__._k(key), value)
+
+    def __delitem__(self, key):
+        return super(CaseInsensitiveDefaultDict, self).__delitem__(self.__class__._k(key))
+
+    def __contains__(self, key):
+        return super(CaseInsensitiveDefaultDict, self).__contains__(
+            self.__class__._k(key)
+        )
+
+    def has_key(self, key):
+        return super(CaseInsensitiveDefaultDict, self).has_key(self.__class__._k(key))
+
+    def pop(self, key, *args, **kwargs):
+        return super(CaseInsensitiveDefaultDict, self).pop(
+            self.__class__._k(key), *args, **kwargs
+        )
```

Comment:
We also need to implement popitem from the mutablemapping api: https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes

Reply:
This seems like a good place to put my comment (thanks Aaron?)

Do we need to implement the full mutable mapping interface? I wonder what this implementation would look like as an abc.Mapping (none of the mutable stuff), that backed on to a dict, rather than extending it. Is there a use case for mutating response.headers?

Reply:
No - in fact, I was going to suggest that it would be better to have a frozen mapping here - but I didn't want to turn this into a Christmas tree of improvement ideas (more than it already is).  But in the abstract -- yes an immutable interface would definitely be preferable.

Reply:
I don't think there is a case for mutating it, not like in the request. I think to make sure this is correct later for use cases of this unrelated to response.headers, we should to more unit tests.